openai/openai.go

```diff
@@ -45,18 +55,20 @@ func (oac *OpenAiClient) Init() {
 	oac.client = openai.NewClient(openaiToken)
 }
 
-func (oac *OpenAiClient) GetCompletion(prompt string) (*string, error) {
+func (oac *OpenAiClient) GetCompletion(prompt string, debug bool) (*string, error) {
 
-	logMsg := "Evaluating code lines:\n"
-	splitLines := strings.Split(prompt, "\n")
-	for i, line := range splitLines {
-		if i < 4 || i > len(splitLines)-4 {
-			logMsg += line + "\n"
-		} else if i == 4 {
-			logMsg += "...\n"
+	if debug {
+		logMsg := "Evaluating code lines:\n"
```

Comment:
Suggestion: The debug logging could be made more specific by including the line numbers being previewed. Consider updating the logging to include this information.
DEBUG INFO
[File] openai/openai.go

[Side] RIGHT

[Start Line] 61

[End Line] 68

[Raw Comment] [+ Line 61-68] Suggestion: The debug logging could be made more specific by including the line numbers being previewed. Consider updating the logging to include this information.
[
[Original Code]  a/openai/openai.go b/openai/openai.go
index 64e969f..d4004ab 100644
 
 const (
 	systemMessage = `
15 -		You are an expert GitHub reviewer. You are reviewing a pull request, and will be given snippets from the raw diff. 
15 +		You are an expert GitHub code reviewer. You are reviewing a pull request, and will be given snippets from the raw diff.
16 +		Lines that start with "-" are lines that have been removed. Lines that start with "+" are lines that have been added.
 		If you see no problems, respond with "No comments".
17 -		If you see any problems, you respond with a comment that references the line number in sqaure brackets and a description of the problem.
18 -		If the problem is a typo, prefix your comment with with "Typo:".
 		If you are certain there is a bug, prefix your comment with "Bug:".
20 -		If the problem is a potential bug, prefix your comment with "Potential Bug:".
21 -		If the problem is a style issue, prefix your comment with "Style:".
19 +		If the problem is a potential bug, do not comment.
20 +		If the problem is a style issue, do not comment.
 		If the problem is a question, prefix your comment with "Question:".
 		If the problem is a suggestion, prefix your comment with "Suggestion:".
 		If the problem is a request for clarification, prefix your comment with "Clarification:".
25 -		If the problem is a request for more information, prefix your comment with "More Info:".
 		If a unit test of critical functionality is missing, prefix your comment with "Missing Test:".
 		If a unit test could use some more test cases, prefix your comment with "Suggested Test Cases:".
26 +		Copy the "+" or "-" into your comment prefix before the line number. 
27 +		Only rarely comment on a line that starts with "-".
28 +		Do not comment on imports.
29 +		Do not nitpick. Comments must be high quality and pithy.
30 +		You can comment on multiple lines.	
 		An example response would look like this:
29 -			"[Line 42] Bug: missing semicolon
30 -			[Line 2] Style: use single quotes instead of double quotes"
31 -		Do not nitpick. Comments must be high quality and pithy. Include code snippets if necessary.	
32 +			[- Line 2] Bug: 'countPizzas' is being used elsewhere and still needs to be initialized
33 +			[+ Line 42] Readability: consider saving this magic number to a variable
34 +			[+ Line 43] Refactor Suggestion: This code is duplicated in 3 places. Consider refactoring into a function.
 	`
 	CODE_PREVIEW_SIZE = 4
37 +	followUpMessage   = `Please make correct any line references you may have gone wrong.
38 +	Remove low quality comments. 
39 +	Improve your answers where appropriate. 
40 +	Remove all comments that have to do with import statements.
41 +	Maintain the same format as your first response.
42 +	If you answered "No comments", respond with "No comments" again.
43 +	If no changes are needed from your last response, respond with "No changes".`
 )
 
 type OpenAiClient struct {
 	oac.client = openai.NewClient(openaiToken)
 }
 
48 -func (oac *OpenAiClient) GetCompletion(prompt string) (*string, error) {
58 +func (oac *OpenAiClient) GetCompletion(prompt string, debug bool) (*string, error) {
 
50 -	logMsg := "Evaluating code lines:\n"
51 -	splitLines := strings.Split(prompt, "\n")
52 -	for i, line := range splitLines {
53 -		if i < 4 || i > len(splitLines)-4 {
54 -			logMsg += line + "\n"
55 -		} else if i == 4 {
56 -			logMsg += "...\n"
60 +	if debug {
61 +		logMsg := "Evaluating code lines:\n"
62 +		splitLines := strings.Split(prompt, "\n")
63 +		for i, line := range splitLines {
64 +			if i < CODE_PREVIEW_SIZE || i > len(splitLines)-CODE_PREVIEW_SIZE {
65 +				logMsg += line + "\n"
66 +			} else if i == CODE_PREVIEW_SIZE {
67 +				logMsg += "...\n"
68 +			}
 		}
70 +		fmt.Print(logMsg)
 	}
59 -	fmt.Print(logMsg)
 
 	resp, err := oac.client.CreateChatCompletion(
 		context.Background(),
 	}
 
 	fmt.Printf(
83 -		"PromptTokens:\t\t%d\nCompletionTokens:\t\t%d\nTotalTokens:\t\t%d",
95 +		"PromptTokens:\t\t%d\nCompletionTokens:\t%d\nTotalTokens:\t\t%d\n",
 		resp.Usage.PromptTokens,
 		resp.Usage.CompletionTokens,
 		resp.Usage.TotalTokens,
 	)
88 -	return &resp.Choices[0].Message.Content, nil
100 +	firstDraft := &resp.Choices[0].Message.Content
101 +
102 +	resp2, err2 := oac.client.CreateChatCompletion(
103 +		context.Background(),
104 +		openai.ChatCompletionRequest{
105 +			Model: openai.GPT3Dot5Turbo,
106 +			Messages: []openai.ChatCompletionMessage{
107 +				{
108 +					Role:    openai.ChatMessageRoleSystem,
109 +					Content: systemMessage,
110 +				},
111 +				{
112 +					Role:    openai.ChatMessageRoleUser,
113 +					Content: prompt,
114 +				},
115 +				{
116 +					Role:    openai.ChatMessageRoleAssistant,
117 +					Content: *firstDraft,
118 +				},
119 +				{
120 +					Role:    openai.ChatMessageRoleUser,
121 +					Content: followUpMessage,
122 +				},
123 +			},
124 +		},
125 +	)
126 +	if err2 != nil {
127 +		return nil, err2
128 +	}
129 +	secondResponse := &resp2.Choices[0].Message.Content
130 +	if strings.Contains(*secondResponse, "No changes") {
131 +		return firstDraft, nil
132 +	}
133 +	return &resp2.Choices[0].Message.Content, nil
 }